### PR TITLE
Add GCC 8 to Travis

### PR DESCRIPTION
**Pull request by tamird**
- Avoid `strdup` to use jemalloc on Windows
- Use `size_t` for consistency
- Add GCC 8 to Travis
- Add CMAKE_BUILD_TYPE=Release to Travis

### CODE REVIEW

**Comment by siying** [comment]
I don't know why the Windows build failed. Can you rebase to relaunch the Windows build?

**Comment by tamird** [comment]
Failure looks legit and it's the same failure that's happening on master:

**Comment by tamird** [comment]
The windows build was broken in https://github.com/facebook/rocksdb/commit/46acdc98838f4545e70751d09be2b507aa90f253 (https://github.com/facebook/rocksdb/pull/3424).

File: memtable/inlineskiplist.h [hunk 1]
```diff
@@ -279,7 +279,11 @@ struct InlineSkipList<Comparator>::Node {
   // next_[0].  This is used for passing data from AllocateKey to Insert.
   void StashHeight(const int height) {
     assert(sizeof(int) <= sizeof(next_[0]));
-    memcpy(&next_[0], &height, sizeof(int));
+    // Explicitly cast to void* because GCC 8 is (rightfully) upset that this is
+    // "writing to an object of type ‘...’ with no trivial copy-assignment".
```
Comment (siying) [inline memtable/inlineskiplist.h]: Our internal lint said, you have non-ascii chars (I assume "‘" and "’") in this line and you should fix it.

Comment (tamird) [reply memtable/inlineskiplist.h]: Ah, copied verbatim from the gcc error message. Will fix tomorrow.

**Comment by siying** [comment]
@tamird also the Windows failure has been fixed in master. When you fix the ASCII-char, please rebase to see whether Windows build passes.

**Comment by tamird** [comment]
Where was it fixed? Still looks broken to me.

File: memtable/inlineskiplist.h (hunk 1 above)
Comment (tamird) [reply memtable/inlineskiplist.h]: Fixed.

**Comment by tamird** [comment]
I'm installing gcc 7.2 locally to try and figure out how to appease [implicit-fallthrough].

**Comment by tamird** [comment]
This is very strange - I'm not able to reproduce this issue with gcc 7.2 on my machine. Rebased, but I imagine it'll fail again.

**Comment by tamird** [comment]
Well, it failed - I'm at a loss. Any ideas?

**Comment by siying** [comment]
@tamird we really appreciate your trying it. Is the problem that you are tired of adding more fall-through hint? If you want to get something quickly in, you can also echo this warning in Makefile if you want.

**Comment by tamird** [comment]
> Is the problem that you are tired of adding more fall-through hint?

No, the problem is that the hint is there, but the warning still shows up in travis (but not locally).

**Comment by miasantreble** [comment]
@tamird I think you are conflicting with the this commit: https://github.com/facebook/rocksdb/commit/e502839e25b224f9488f46da2df9d896ead71403#diff-f54be34b9b64b37988070860edd87aaf
Can you fetch the latest upstream and rebase again? Thanks

**Comment by tamird** [comment]
Rebased. This is still exhibiting the same problem:
```
db/log_reader.cc:150:9: error: this statement may fall through [-Werror=implicit-fallthrough=]
         if (wal_recovery_mode == WALRecoveryMode::kAbsoluteConsistency) {
         ^~
db/log_reader.cc:156:7: note: here
       case kEof:
       ^~~~
db/log_reader.cc:170:9: error: this statement may fall through [-Werror=implicit-fallthrough=]
         if (wal_recovery_mode != WALRecoveryMode::kSkipAnyCorruptedRecords) {
         ^~
db/log_reader.cc:186:7: note: here
       case kBadRecord:
       ^~~~
```
which makes no sense, since both of those `case`s have `// FALLTHRU` comments.

**Comment by miasantreble** [comment]
I'm no expert on gcc but according to [this post](https://stackoverflow.com/questions/45129741/gcc-7-wimplicit-fallthrough-warnings-and-portable-way-to-clear-them) you need to remove the empty line between the comment and the actual code?

**Comment by tamird** [comment]
@miasantreble doesn't look like that helped :(

**Comment by tamird** [comment]
This is still failing, but I can reproduce it now! Working on it.

**Comment by yiwu-arbug** [comment]
@tamird Appreciate your contribution. Can you rebase and fix travis? Seems there's more compile errors. Appveyor error is due to flaky test and unrelated.

File: .travis.yml [hunk 2]
```diff
@@ -15,22 +15,32 @@ cache:
...
 
 addons:
-   apt:
-      packages: ['zlib1g-dev', 'libbz2-dev', 'libsnappy-dev', 'curl', 'libgflags-dev', 'mingw-w64']
+  apt:
+    sources:
+      - ubuntu-toolchain-r-test
+    packages:
+      - curl
+      - g++-7
```
Comment (yiwu-arbug) [inline .travis.yml]: can we existing tests compile with gcc4.8, and add a build to compile with gcc7? And instead of gcc7, shall we do gcc8?

**Comment by itamarjp** [comment]
one alternative to travis is using fedora copr, they allways has the newer version of the compilers, 

https://copr.fedorainfracloud.org

we can grab the spec file used by fedora, and a github webhook will trigger the build after a commit.

File: table/table_test.cc [hunk 3]
```diff
@@ -977,14 +977,14 @@ class HarnessTest : public testing::Test {
...
-    for (unsigned int i = static_cast<unsigned int>(start_i); i < end_i; i++) {
+    for (size_t i = start_i; i < end_i; i++) {
       Init(args[i]);
       Random rnd(test::RandomSeed() + 5);
       for (int num_entries = 0; num_entries < 2000;
            num_entries += (num_entries < 50 ? 1 : 200)) {
         if ((num_entries % 10) == 0) {
-          fprintf(stderr, "case %d of %d: num_entries = %d\n", (i + 1),
-                  static_cast<int>(args.size()), num_entries);
+          fprintf(stderr, "case %zu of %zu: num_entries = %d\n", (i + 1),
```
Comment (yiwu-arbug) [inline table/table_test.cc]: use ROCKSDB_PRIszt macro instead of "%zu"?

File: .travis.yml (hunk 2 above)
Comment (yiwu-arbug) [reply .travis.yml]: We want to keep rocksdb compatible with gcc4.8. How about we just add one job to compile with gcc7?

Comment (tamird) [reply .travis.yml]: Done. That job still fails, though.

**Comment by tamird** [comment]
Wow, this same fallthrough problem persists even with gcc 8 (but does not reproduce on my machine). I'll have to find some time to spin up a docker container with Travis' image to try and figure this out.

**Comment by tamird** [comment]
@yiwu-arbug I've finally worked around the implicit fallthrough issue - PTAL.

There was a test failure - it looks unrelated.

**Comment by tamird** [comment]
@siying ping

**Comment by miasantreble** [comment]
LGTM, restarted the failed travis job.

**Comment by tamird** [comment]
Anything left holding this up?

**Pull request closed without merging**